File: rtproc/Python/PortAudioPipe.py
# The code in this file is ported from the C++ code originally written by
# Paolo Prandoni and Martin Vetterli of the Ecole Polytechnique Federale
# de Lausanne.

import pyaudio
import select
import sys
import os
import struct
import wave
import SoundProcessor

FRAMES_PER_BUFFER = 4
NUM_CHANNELS = 1
SAMPLE_SIZE = 2
SAMPLE_RATE = 44100
PA_SAMPLE_TYPE = pyaudio.paInt32

effects = {
            "0" : "Delta",
            "1" : "Echo",
            "2" : "IIR echo",
            "3" : "Natural Echo",
            "4" : "Reverb",
            "5" : "Biquad",
            "6" : "Fuzz",
            "7" : "Flanger",
            "8" : "Wah",
            "9" : "Tremolo"
           }

class PortAudioPipe:
    def __init__(self, sound_file=None):
        self._sound_file = None
        self._pa = None
        self._istream = None
        self._ostream = None
        self._sample_block = 0.0
        self._sound_processor = None
        self._effect = None
        self._err = 0
        self._sound_file = sound_file
        self._frames_per_buffer = FRAMES_PER_BUFFER
 
    def initialize(self):
        # Initialize the PortAudio library
        self._pa = pyaudio.PyAudio()
        
        # If we have a sound file, set up the parameters here 
        if not self._sound_file == None:
            self._istream = wave.open(self._sound_file, 'rb')
            self._sample_rate = self._istream.getframerate()
            self._sample_size = 1
            self._frames_per_buffer = 1
            self._num_channels = self._istream.getnchannels()
            self._format = self._pa.get_format_from_width(self._istream.getsampwidth())
            if self._format == pyaudio.paUInt8:
                self._pack_format = 'B'
            elif self._format == pyaudio.paInt8:
                self._pack_format = 'b'                
            elif self._format == pyaudio.paInt16:
                self._pack_format = 'h'
            elif self._format == pyaudio.paInt24:
                # TODO
                pass
            elif self._format == pyaudio.paInt32:
                self._pack_format = 'i'
            elif self._format == pyaudio.paFloat32:
                self._pack_foramt = 'f'
        else:   
            self._sample_rate = 44100
            self._sample_size = 2
            self._frames_per_buffer = 4
            self._num_channels = 1
            self._format = pyaudio.paInt32
            self._pack_format = 'i'     
            self._istream = self._pa.open(
                                    channels=self._num_channels,
                                    format=self._format,
                                    rate=self._sample_rate,
                                    frames_per_buffer=self._frames_per_buffer,
                                    input=True)  
                        
        self._ostream = self._pa.open(
                            channels=self._num_channels,
                            format=self._format,
                            rate=self._sample_rate,
                            frames_per_buffer=self._frames_per_buffer,
                            output=True) 
        self._sample_block = [0.0] * self._frames_per_buffer * self._num_channels * self._sample_size
 
        # Set up the SoundProcessor
        self._sound_processor = SoundProcessor.SoundProcessor(SR=self._sample_rate)

    def _play_sample(self):
        try:
            if not self._sound_file == None:
                sample_block = self._istream.readframes(self._frames_per_buffer)
            else:
                sample_block = self._istream.read(self._frames_per_buffer)
            current_block = []
            for j in range(0,len(sample_block),self._frames_per_buffer):
                aux = long(struct.unpack(self._pack_format, sample_block[j:j+self._frames_per_buffer])[0])
                current_block = current_block + \
                    [struct.pack(self._pack_format, self._sound_processor.Process(float(aux)))]
            self._ostream.write(''.join(current_block), self._frames_per_buffer)
        except IOError as e:
            self._err = e.errno
            self.xrun()
        except struct.error:
            # Occasionally, a sample goes out of range for the struct.pack
            # function.  Rather than fail out, allow the processing to
            # continue, effectively throwing this sample out
            pass
        
    def start(self):        
        legal = False
        while not legal:
            self._print_options()
            effect = raw_input()
            if effect in effects.keys():
                self._effect = effect
                self._sound_processor.SetFunction(effect)
                self._print_options()
                legal = True
            else:
                print
                print "Please make a legal selection..."
                print
                print
                
        # Start the streams
        if self._sound_file == None:
            self._istream.start_stream()
        self._ostream.start_stream()
        
        # Loop until we're told to quit
        while True:
            key, _, _ = select.select([sys.stdin], [], [], 0)
            if key == []:
                self._play_sample()
            else:   
                effect = sys.stdin.read(1)
                if effect.upper() == 'Q':
                    print "Thanks for playing!"
                    self.stop()
                    break
                if effect in effects.keys():
                    self._effect = effect
                    self._sound_processor.SetFunction(effect)
                    self._print_options()
                    if not self._sound_file == None:
                        self._istream.rewind()
        
    def stop(self):
        if not self._istream == None:
            if self._sound_file == None:
                self._istream.stop_stream()
            self._istream.close()
        if not self._ostream == None:
            self._ostream.stop_stream()
            self._ostream.close()            
        if not self._pa == None:
            self._pa.terminate()
        
    def _print_options(self):
        def print_selector():
            print '[*]',

        os.system('cls' if os.name == 'nt' else 'clear')

        if self._effect == '0': print_selector()
        print "(0) " + effects['0']
        if self._effect == '1': print_selector()
        print "(1) " + effects['1']
        if self._effect == '2': print_selector()
        print "(2) " + effects['2']
        if self._effect == '3': print_selector()
        print "(3) " + effects['3']
        if self._effect == '4': print_selector()
        print "(4) " + effects['4']
        if self._effect == '5': print_selector()
        print "(5) " + effects['5']
        if self._effect == '6': print_selector()
        print "(6) " + effects['6']
        if self._effect == '7': print_selector()
        print "(7) " + effects['7']
        if self._effect == '8': print_selector()
        print "(8) " + effects['8']
        if self._effect == '9': print_selector()
        print "(9) " + effects['9']
        print "(Q) Quit"
        print
        print "Please make a selection: "
        
    def xrun(self):
        self.stop()
        if self._err == pyaudio.paInputOverflow:
            print "Input Overflow."
        if self._err == pyaudio.paOutputUnderflow:
            print "output Underflow."

File: rtproc/C++/portAudioPipe.cpp
// Created by Paolo Prandoni and Martin Vetterli of the Ecole Polytechnique
// Federale de Lausanne

#include "portAudioPipe.h"

#ifndef WIN32
int _kbhit(void) {
    int ch;
    wtimeout(stdscr, 0);
    ch = wgetch(stdscr);
    nodelay(stdscr, FALSE);
    if (ch == ERR) return 0;
    ungetch(ch);
    return 1;    
}
#endif

portAudioPipe::portAudioPipe()
{		
}

portAudioPipe::~portAudioPipe()
{
}

//Intialize all the component
void portAudioPipe:: Initial()
{
    //create the Buffer
    int numBytes = FRAMES_PER_BUFFER * NUM_CHANNELS * SAMPLE_SIZE ;
  
    sampleBlock = (char *) malloc( numBytes );
    if( sampleBlock == NULL )
    {
        printf("Could not allocate record array.\n");
        exit(1);
    }
    CLEAR( sampleBlock );	
  
    //Initialize the PortAudio library 
    err = Pa_Initialize();
    if( err != paNoError ) 
        error();
	
    //Initialize the input device
    inputParameters.device = Pa_GetDefaultInputDevice(); /* default input device */
    inputParameters.channelCount = NUM_CHANNELS;
    inputParameters.sampleFormat = PA_SAMPLE_TYPE;
    inputParameters.suggestedLatency = Pa_GetDeviceInfo( inputParameters.device )->defaultHighInputLatency ;
    inputParameters.hostApiSpecificStreamInfo = NULL;

    //Initialize the output device
    outputParameters.device = Pa_GetDefaultOutputDevice(); /* default output device */
    outputParameters.channelCount = NUM_CHANNELS;
    outputParameters.sampleFormat = PA_SAMPLE_TYPE;
    outputParameters.suggestedLatency = Pa_GetDeviceInfo( outputParameters.device )->defaultHighOutputLatency;
    outputParameters.hostApiSpecificStreamInfo = NULL;

    //Intialize soundProcessor with Sample rate
    m_soundProcessor.Init(SAMPLE_RATE);

    initscr();
}

void portAudioPipe::Start()
{
    int r;

    //Display all possible sound effect and get the first sound effect
    printOptions(0);
    m_soundProcessor.SetFunction(option);

    //start the stream
    err = Pa_OpenStream(
                &stream,
                &inputParameters,
                &outputParameters,
                SAMPLE_RATE,
                FRAMES_PER_BUFFER,
                paClipOff,          /* we won't output out of range samples so don't bother clipping them */
                NULL,               /* no callback, use blocking API */
                NULL );             /* no callback, so no callback userData */
  
    if( err != paNoError ) 
        error();

    err = Pa_StartStream( stream );
    if( err != paNoError ) 
        error();

    //Infinite loop for all the operations
    while(1)
    {
        r = _kbhit();
        if(r == 0)    // no key is pressed on the keyboard
        {
            err = Pa_ReadStream( stream, sampleBlock, FRAMES_PER_BUFFER );
            if( err && CHECK_OVERFLOW )
                xrun();       
	   
            currentBlock = sampleBlock;
            //read samples from the buffer and put them back after operations
            for(int j = 0; j < FRAMES_PER_BUFFER; j++)
            {
                int tmp = *((__int16*)(currentBlock));
                *((__int16*)(currentBlock)) = (__int16)m_soundProcessor.Process((float)tmp);
                currentBlock+=2;    //+2 because the size of the data is two byte
            }
	   
            err = Pa_WriteStream( stream, sampleBlock, FRAMES_PER_BUFFER );
            if( err && CHECK_UNDERFLOW ) 
                xrun();       
        }
        else   // If user press the keyboard    
        {
            int option = _getch();
            if (option == 'q' || option == 'Q')
            {
                Stop();
                return;
            }
            option -= '0';
            m_soundProcessor.SetFunction(option);
            printOptions(option);
        } 
    }
}

void portAudioPipe::Stop()
{
    if( stream ) 
    {
        Pa_AbortStream( stream );
        Pa_CloseStream( stream );
    }
    free( sampleBlock );
    Pa_Terminate();
    endwin();

    return;
}

void portAudioPipe::printOptions(int option)
{
    clear();
    PRINTOPTION("#####################################\n");
    PRINTOPTION("Please choose a sound effect:\n");
    if (option == 0)
        PRINTOPTION("[*] ");
    PRINTOPTION("0: delta\n");
    if (option == 1)
        PRINTOPTION("[*] ");
    PRINTOPTION("1: echo\n");
    if (option == 2)
        PRINTOPTION("[*] ");
    PRINTOPTION("2: IIR echo\n");
    if (option == 3)
        PRINTOPTION("[*] ");
    PRINTOPTION("3: natural echo\n");
    if (option == 4)
        PRINTOPTION("[*] ");
    PRINTOPTION("4: reverb\n");
    if (option == 5)
        PRINTOPTION("[*] ");
    PRINTOPTION("5: biquad\n");
    if (option == 6)
        PRINTOPTION("[*] ");
    PRINTOPTION("6: fuzz\n");
    if (option == 7)
        PRINTOPTION("[*] ");
    PRINTOPTION("7: flanger\n");
    if (option == 8)
        PRINTOPTION("[*] ");
    PRINTOPTION("8: wah\n");
    if (option == 9)
        PRINTOPTION("[*] ");
    PRINTOPTION("9: tremolo\n");
    PRINTOPTION("\n");
    PRINTOPTION("Q: quit\n");
}

void portAudioPipe::error()
{
    if( stream ) 
    {
        Pa_AbortStream( stream );
        Pa_CloseStream( stream );
    }

    free( sampleBlock );
    Pa_Terminate();
    fprintf( stderr, "An error occured while using the portaudio stream\n" );
    fprintf( stderr, "Error number: %d\n", err );
    fprintf( stderr, "Error message: %s\n", Pa_GetErrorText( err ) );
    return;
}


void portAudioPipe::xrun()
{
    if( stream ) 
    {
        Pa_AbortStream( stream );
        Pa_CloseStream( stream );
    }
    free( sampleBlock );
    Pa_Terminate();
    if( err & paInputOverflow )
        fprintf( stderr, "Input Overflow.\n" );
    if( err & paOutputUnderflow )
        fprintf( stderr, "Output Underflow.\n" );
    return;
}

